File: app/models/snippet.rb
class Snippet < ApplicationRecord
  include CommonApi

  ARTICLE_MINI_PREVIEW_TAG = 'Article'
  GITHUB_URL_PREFIX = 'https://github.com/Chalarangelo/30-seconds-of-code/blob/master/content/snippets'

  # Relationships
  has_one :language, primary_key: 'language_cid', foreign_key: 'cid'

  has_many :collection_snippets,
    class_name: 'CollectionSnippet',
    foreign_key: 'snippet_cid'

  has_many :collections,
    through: :collection_snippets

  # Prevent certain attributes from getting printed in the console
  self.filter_attributes += [
    :short_description,
    :description,
    :table_of_contents
  ]

  # TODO: Rails 7.2 will introduce this, check later
  # self.attributes_for_inspect = [:id, :cid]

  scope :by_new, -> { order('date_modified desc') }

  scope :unlisted, -> { where(listed: false) }
  scope :listed, -> { where(listed: true) }

  # We explicitly name the model here to avoid conflicts with the
  # `CollectionSnippet` model having a column of the same name.
  scope :published, -> do
     where(snippets: { date_modified: ..Date.today })
  end
  scope :scheduled, -> do
    where.not(snippets: { date_modified: ..Date.today })
  end

  scope :previewable, -> { all.preload(:language) }

  def tags
    @tags ||= _tags.split(';')
  end

  def has_language?
    @has_language ||= language_cid.present?
  end

  def seo_title
    return title unless has_language?

    title_language =
      if language_cid == 'javascript' && primary_tag === 'node'
        formatted_primary_tag
      else
        language.name
      end

    title.include?(title_language) ? title : "#{title_language} - #{title}"
  end

  def primary_tag
    @primary_tag ||= tags.first
  end

  def formatted_primary_tag
    TagFormatter.format(primary_tag)
  end

  # Used for snippet previews in search autocomplete
  def formatted_mini_preview_tag
    @formatted_mini_preview_tag ||=
      if has_language?
        language.name
      else
        ARTICLE_MINI_PREVIEW_TAG
      end
  end

  def formatted_tags
    return @formatted_tags if defined?(@formatted_tags)

    @formatted_tags = tags.map { |tag| TagFormatter.format(tag) }
    @formatted_tags.prepend(language.name) if has_language?
    @formatted_tags = @formatted_tags.join(', ')
  end

  def formatted_preview_tags
    @formatted_preview_tags ||=
      if has_language?
        language.name
      else
        formatted_primary_tag
      end
  end

  def github_url
    @github_url ||= "#{GITHUB_URL_PREFIX}#{slug}.md"
  end

  def is_scheduled?
    @is_scheduled ||= date_modified > Date.today
  end

  def is_published?
    @is_published ||= !is_scheduled?
  end

  def is_listed?
    @is_listed ||= listed && is_published?
  end

  def date_formatted
    @date_formatted ||= date_modified.strftime('%B %-d, %Y')
  end

  def date_machine_formatted
    @date_machine_formatted ||= date_modified.strftime('%Y-%m-%d')
  end

  def slug_id
    @slug_id ||= slug.split('/').last
  end

  def search_tokens_array
    @search_tokens_array ||= [
      slug_id,
      *tags,
      *_tokens.split(';'),
      *title.normalized_tokens,
      language&.short&.downcase,
      language&.long&.downcase,
    ].compact.uniq
  end

  def has_collection?
    collection_snippets.present?
  end

  def breadcrumbs
    @breadcrumbs ||= breadcrumbs_presenter.breadcrumbs
  end

  def recommended_collection
    @recommended_collection ||= breadcrumbs_presenter.recommended_collection
  end

  def recommended_snippets
    @recommended_snippets ||= recommendation_presenter.recommend_snippets
  end

  def page
    @page ||= Page.from(self)
  end

  private

  def breadcrumbs_presenter
    @breadcrumbs_presenter ||= BreadcrumbPresenter.new(self)
  end

  def recommendation_presenter
    @recommendation_presenter ||= RecommendationPresenter.new(self)
  end
end

File: src/models/snippet.js
import Model from '#src/core/model.js';
import ContentModel from '#src/models/contentModel.js';
import Language from '#src/models/language.js';
import CollectionSnippet from '#src/models/collectionSnippet.js';
import Collection from '#src/models/collection.js';
import BreadcrumbPresenter from '#src/presenters/breadcrumbPresenter.js';
import RecommendationPresenter from '#src/presenters/recommendationPresenter.js';
import Page from '#src/adapters/page.js';
import StringUtils from '#src/lib/stringUtils.js';
import settings from '#src/config/settings.js';

export default class Snippet extends ContentModel {
  static {
    Model.prepare(this, ['id']);
  }

  constructor(data) {
    super(data);
    this.id = data.id;
    this.title = data.title;
    this.shortTitle = data.shortTitle;
    this.content = data.content;
    this.description = data.description;
    this.listed = data.listed || false;
    this.cover = data.cover;
    this.tokens = data.tokens.split(';');
    this.docTokens = data.docTokens.split(';');
    this.ranking = data.ranking;
    this.tags = data.tags.split(';');
    this.dateModified = new Date(data.dateModified);
    this.tableOfContents = data.tableOfContents;
    this.languageId = data.languageId;
    this.journeyId = data.journeyId;
  }

  static byNew(records) {
    return records.order((a, b) => b.dateModified - a.dateModified);
  }

  static last30Days(records) {
    let date = new Date();
    date.setDate(date.getDate() - 30);
    return records.where({ dateModified: d => d > date });
  }

  static published(records) {
    const now = new Date();
    return records.where({ dateModified: d => d < now });
  }

  static scheduled(records) {
    const now = new Date();
    return records.where({ dateModified: d => d > now });
  }

  static updateLogs(records) {
    return records.where({
      tags: t => t.includes(settings.collections.updateLogTag),
    });
  }

  static inJourney(records) {
    return records.where({ journeyId: j => j !== null });
  }

  get language() {
    return Language.find(this.languageId);
  }

  get collectionSnippets() {
    return CollectionSnippet.where({ snippetId: this.id });
  }

  get collections() {
    return Collection.where({
      id: this.collectionSnippets.pluck('collectionId'),
    });
  }

  get journey() {
    if (!this.journeyId) return null;
    return Collection.find(this.journeyId);
  }

  get previousJourneySnippet() {
    if (!this.journey) return null;

    const snippets = this.journey.listedSnippets;
    const index = snippets.findIndex(s => s.id === this.id);
    return snippets[index - 1] || null;
  }

  get nextJourneySnippet() {
    if (!this.journey) return null;

    const snippets = this.journey.listedSnippets;
    const index = snippets.findIndex(s => s.id === this.id);
    return snippets[index + 1] || null;
  }

  get journeyPagination() {
    if (!this.journey) return null;

    return {
      previousUrl: this.previousJourneySnippet?.url,
      nextUrl: this.nextJourneySnippet?.url,
      url: this.journey.url,
      title: this.journey.miniTitle,
      totalItems: this.journey.listedSnippets.length,
      itemNumber:
        this.journey.listedSnippets.findIndex(s => s.id === this.id) + 1,
    };
  }

  get hasLanguage() {
    return Boolean(this.language);
  }

  get seoTitle() {
    if (!this.hasLanguage) return this.title;

    const titleLanguage =
      this.languageId === 'javascript' && this.primaryTag === 'node'
        ? this.formattedPrimaryTag
        : this.language.name;

    return this.title.includes(titleLanguage)
      ? this.title
      : `${titleLanguage} - ${this.title}`;
  }

  get primaryTag() {
    return this.tags[0];
  }

  get formattedPrimaryTag() {
    return StringUtils.formatTag(this.primaryTag);
  }

  // Used for snippet previews in search autocomplete
  get formattedMiniPreviewTag() {
    if (this.hasLanguage) return this.language.name;
    return settings.presentation.articlePreviewTag;
  }

  get formattedTags() {
    const formattedTags = this.tags.map(tag => StringUtils.formatTag(tag));
    if (this.hasLanguage) formattedTags.unshift(this.language.name);
    return formattedTags.join(', ');
  }

  get formattedPreviewTags() {
    if (this.hasLanguage) return this.language.name;
    return this.formattedPrimaryTag;
  }

  get githubUrl() {
    return `${settings.repository.snippetPrefix}${this.slug}.md`;
  }

  get isScheduled() {
    return this.dateModified > new Date();
  }

  get isPublished() {
    return !this.isScheduled;
  }

  get isListed() {
    return this.listed && this.isPublished;
  }

  get isUpdateLog() {
    return this.tags.includes(settings.collections.updateLogTag);
  }

  get dateFormatted() {
    return this.dateModified.toLocaleDateString('en-US', {
      day: 'numeric',
      month: 'long',
      year: 'numeric',
    });
  }

  get dateMachineFormatted() {
    return this.dateModified.toISOString().slice(0, 10);
  }

  get dateShortString() {
    return this.dateModified.toISOString().split('T')[0];
  }

  get searchTokensArray() {
    return [
      ...new Set(
        [
          this.slugId,
          ...this.tags,
          ...this.tokens,
          ...StringUtils.normalizedTokens(this.title),
          this.language?.short?.toLowerCase(),
          this.language?.long?.toLowerCase(),
        ].filter(Boolean)
      ),
    ];
  }

  // TODO: This returns true even if the snippet only belongs to the
  // main collection. The `CollectionSnippet` model must be updated
  // with a method that checks for the main collection (via id).
  get hasCollection() {
    return this.collectionSnippets.length > 0;
  }

  get breadcrumbs() {
    return this.breadcrumbsPresenter.breadcrumbs;
  }

  get recommendedCollection() {
    return this.breadcrumbsPresenter.recommendedCollection;
  }

  get recommendedSnippets() {
    return this.recommendationPresenter.recommendSnippets();
  }

  get page() {
    return Page.from(this);
  }

  get breadcrumbsPresenter() {
    return new BreadcrumbPresenter(this);
  }

  get recommendationPresenter() {
    return new RecommendationPresenter(this);
  }
}
